cpp/src/utilities/type_checks.cpp

```diff
@@ -46,33 +49,114 @@ bool columns_equal_fn::operator()<dictionary32>(column_view const& lhs, column_v
 {
+  if (rhs.type().id() != type_id::STRUCT) { return false; }
   return lhs.num_children() == rhs.num_children() and
          std::all_of(thrust::make_counting_iterator(0),
                      thrust::make_counting_iterator(lhs.num_children()),
                      [&](auto i) { return column_types_equal(lhs.child(i), rhs.child(i)); });
 }
 
+struct column_scalar_equal_fn {
+  template <typename T>
+  bool operator()(column_view const& col, scalar const& slr)
+  {
+    return col.type() == slr.type();
+  }
+};
+
+template <>
+bool column_scalar_equal_fn::operator()<dictionary32>(column_view const& col, scalar const& slr)
+{
+  // It is not possible to have a scalar dictionary, so compare the dictionary
+  // column keys type to the scalar type.
+  auto col_keys = cudf::dictionary_column_view(col).keys();
+  return column_scalar_types_equal(col_keys, slr);
+}
+
+template <>
+bool column_scalar_equal_fn::operator()<list_view>(column_view const& col, scalar const& slr)
+{
+  if (slr.type().id() != type_id::LIST) { return false; }
+  auto const& ci      = lists_column_view::child_column_index;
+  auto const list_slr = static_cast<list_scalar const*>(&slr);
+  return column_types_equal(col.child(ci), list_slr->view());
+}
+
+template <>
+bool column_scalar_equal_fn::operator()<struct_view>(column_view const& col, scalar const& slr)
+{
+  if (slr.type().id() != type_id::STRUCT) { return false; }
+  auto const struct_slr = static_cast<struct_scalar const*>(&slr);
+  auto const slr_tbl    = struct_slr->view();
+  return col.num_children() == slr_tbl.num_columns() and
+         std::all_of(thrust::make_counting_iterator(0),
+                     thrust::make_counting_iterator(col.num_children()),
+                     [&](auto i) { return column_types_equal(col.child(i), slr_tbl.column(i)); });
+}
+
+struct scalars_equal_fn {
+  template <typename T>
+  bool operator()(scalar const& lhs, scalar const& rhs)
+  {
+    return lhs.type() == rhs.type();
+  }
+};
+
+template <>
+bool scalars_equal_fn::operator()<list_view>(scalar const& lhs, scalar const& rhs)
+{
+  if (rhs.type().id() != type_id::LIST) { return false; }
+  auto const list_lhs = static_cast<list_scalar const*>(&lhs);
+  auto const list_rhs = static_cast<list_scalar const*>(&rhs);
+  return column_types_equal(list_lhs->view(), list_rhs->view());
+}
+
+template <>
+bool scalars_equal_fn::operator()<struct_view>(scalar const& lhs, scalar const& rhs)
+{
+  if (rhs.type().id() != type_id::STRUCT) { return false; }
+  auto const tbl_lhs = static_cast<struct_scalar const*>(&lhs)->view();
+  auto const tbl_rhs = static_cast<struct_scalar const*>(&rhs)->view();
+  return tbl_lhs.num_columns() == tbl_rhs.num_columns() and
+         std::all_of(
+           thrust::make_counting_iterator(0),
+           thrust::make_counting_iterator(tbl_lhs.num_columns()),
+           [&](auto i) { return column_types_equal(tbl_lhs.column(i), tbl_rhs.column(i)); });
+}
```

Comment:
Can we actually iterate over the children of a column?

Then this would be
```suggestion
   return std::equal(tbl_lhs.begin(), tbl_lhs.end(), tbl_rhs.begin(), tbl_rhs.end(),
          [](const auto& lhs, const auto& rhs) { return column_types_equal(rhs, lhs); });
```

Reply:
I think this works, since `table_view.begin()` returns an iterator over the columns.

Reply:
Great idea. This is done.